bftengine/src/bcstatetransfer/BCStateTran.cpp

```diff
@@ -2736,6 +2736,10 @@ void BCStateTran::processData(bool lastInBatch) {
 
       g.txn()->setIsFetchingState(false);
       ConcordAssertEQ(getFetchingState(), FetchingState::NotFetching);
+      // Note: Key exchange message is sent as internal bft client request and
+      // it will be dropped by message handler if curr_state == fetching
+      // So we need to invoke this cb after setting state to NotFetching
+      replicaForStateTransfer_->checkForKeyExchange();
```

Comment:
Instead of defining a virtual method in IState transfer, you can simply register a callback to the state transfer call back registry (on BCState transfer)
See https://github.com/vmware/concord-bft/blob/master/bftengine/src/bcstatetransfer/BCStateTran.cpp#L2730

Reply:
The reason i did not register a callback with BCStateTansfer::onTransferringComplete because callbacks are invoked first and ST fetching state is changed later to NOT_FETCHING. 
KeyExchangeManager::sendKeyExchange sends an internal bft client request which non-primary replica forwards to primary for consensus. 
Problem is that, this internal client request msg will get dropped by ReplicaImp::msgHandler because ST fetching state is FETCHING when onTransferComplete() invokes the callbacks. Please check this condition here:
https://github.com/vmware/concord-bft/blob/master/bftengine/src/bftengine/ReplicaImp.cpp#:~:text=%7D-,if%20(!isCollectingState())%20%7B,-//%20The%20message%20validation 

Please let me know if this makes sense

Reply:
correct, you are right.
But wouldn't it be better to have another registry for callbacks after the state == not fetching?
I think that it will simplify the MR as well as create a good infra for future callbacks of this kind.
WDYT?

Reply:
yes, thats a good point. Let me update the PR as per this. Thanks